dev/src/v1/firestore_client.ts

```diff
@@ -0,0 +1,1318 @@
+    optionsOrCallback?:
+      | gax.CallOptions
+      | Callback<
+          protosTypes.google.firestore.v1.IDocument,
+          protosTypes.google.firestore.v1.IGetDocumentRequest | undefined,
+          {} | undefined
+        >,
+    callback?: Callback<
+      protosTypes.google.firestore.v1.IDocument,
+      protosTypes.google.firestore.v1.IGetDocumentRequest | undefined,
+      {} | undefined
+    >
+  ): Promise<
+    [
+      protosTypes.google.firestore.v1.IDocument,
+      protosTypes.google.firestore.v1.IGetDocumentRequest | undefined,
+      {} | undefined
+    ]
+  > | void {
+    request = request || {};
+    let options: gax.CallOptions;
+    if (typeof optionsOrCallback === 'function' && callback === undefined) {
+      callback = optionsOrCallback;
+      options = {};
+    } else {
+      options = optionsOrCallback as gax.CallOptions;
+    }
+    options = options || {};
+    options.otherArgs = options.otherArgs || {};
+    options.otherArgs.headers = options.otherArgs.headers || {};
+    options.otherArgs.headers[
+      'x-goog-request-params'
+    ] = gax.routingHeader.fromParams({
+      name: request.name || '',
+    });
+    return this._innerApiCalls.getDocument(request, options, callback);
+  }
+  createDocument(
+    request: protosTypes.google.firestore.v1.ICreateDocumentRequest,
+    options?: gax.CallOptions
+  ): Promise<
+    [
+      protosTypes.google.firestore.v1.IDocument,
+      protosTypes.google.firestore.v1.ICreateDocumentRequest | undefined,
+      {} | undefined
+    ]
+  >;
+  createDocument(
+    request: protosTypes.google.firestore.v1.ICreateDocumentRequest,
+    options: gax.CallOptions,
+    callback: Callback<
+      protosTypes.google.firestore.v1.IDocument,
+      protosTypes.google.firestore.v1.ICreateDocumentRequest | undefined,
+      {} | undefined
+    >
+  ): void;
+  /**
+   * Creates a new document.
+   *
+   * @param {Object} request
+   *   The request object that will be sent.
+   * @param {string} request.parent
+   *   Required. The parent resource. For example:
+   *   `projects/{project_id}/databases/{database_id}/documents` or
+   *   `projects/{project_id}/databases/{database_id}/documents/chatrooms/{chatroom_id}`
+   * @param {string} request.collectionId
+   *   Required. The collection ID, relative to `parent`, to list. For example: `chatrooms`.
+   * @param {string} request.documentId
+   *   The client-assigned document ID to use for this document.
+   *
+   *   Optional. If not specified, an ID will be assigned by the service.
+   * @param {google.firestore.v1.Document} request.document
+   *   Required. The document to create. `name` must not be set.
+   * @param {google.firestore.v1.DocumentMask} request.mask
+   *   The fields to return. If not set, returns all fields.
+   *
+   *   If the document has a field that is not present in this mask, that field
+   *   will not be returned in the response.
+   * @param {object} [options]
```

Comment:
optionsOrCallback?

Should the documentation be on the "overload" rather than the catch-all implementation?

Also... uhm... it's 2020 soon. We should drop callbacks :)

Reply:
It's 2020 indeed, and TypeScript still cannot process documentation on overloads :)  If we move the comment block a few lines up so it was before the overloads, it gets completely lost when transpiling to JS, and jsdoc does not get anything at all. Unfortunately, we still use jsdoc on the compiled code to build HTML docs, we haven't found any better solution for TypeScript so far.

As for callbacks - we do advertise `async`/`await` but we still accept callbacks, and dropping them would be a semver major. Actually, dropping callbacks altogether might be a good idea for this January (together with Node 8 deprecation).